scripts/utils/train_helper.py

```diff
@@ -1,131 +1,258 @@
                                          unallowed_types = TeamType.ALL_TYPES_BESIDES_SP)
 
+
+    if TeamType.SELF_PLAY_ADVERSARY in n_x_sp_train_types:
+        prefix = 'PWADV' + '-N-' + str(unseen_teammates_len) + '-SP'
+        suffix = args.primary_learner_type + f'_attack{attack_rounds-1}'
+    else:
+        prefix = 'N-' + str(unseen_teammates_len) + '-SP'
+        suffix = args.primary_learner_type
+
     name = generate_name(args,
-                         prefix = f'N-{args.unseen_teammates_len}-SP',
+                         prefix = prefix,
                          seed = args.N_X_SP_seed,
-                         h_dim = args.N_X_SP_h_dim, 
+                         h_dim = args.N_X_SP_h_dim,
                          train_types = n_x_sp_train_types,
-                         has_curriculum = not curriculum.is_random)
-    
-    agents = load_agents(args, name=name, tag=tag, force_training=n_x_sp_force_training)
+                         has_curriculum = not curriculum.is_random,
+                         suffix=suffix,
+                         )
+    agents = load_agents(args, name=name, tag=tag, force_training=args.primary_force_training)
     if agents:
         return agents[0]
-    
+
     population = get_population(
         args=args,
-        ck_rate=pop_total_training_timesteps // 20,
-        total_training_timesteps=pop_total_training_timesteps,
+        ck_rate=args.pop_total_training_timesteps // 20,
+        total_training_timesteps=args.pop_total_training_timesteps,
         train_types=n_x_sp_train_types,
         eval_types=n_x_sp_eval_types['generate'],
-        unseen_teammates_len = args.unseen_teammates_len,
-        num_SPs_to_train=num_SPs_to_train,
-        parallel=parallel,
-        force_training=pop_force_training,
+        unseen_teammates_len = unseen_teammates_len,
+        num_SPs_to_train=args.num_SPs_to_train,
+        force_training=args.pop_force_training,
         tag = tag
     )
 
-    randomly_init_agent = RLAgentTrainer.generate_randomly_initialized_agent(args=args, seed=args.N_X_SP_seed)
+    
+    if TeamType.SELF_PLAY_ADVERSARY in n_x_sp_train_types:
+        attack_N_X_SP(args=args,
+                      population=population,
+                      curriculum=curriculum,
+                      unseen_teammates_len=unseen_teammates_len,
+                      adversary_play_config=adversary_play_config,
+                      attack_rounds=attack_rounds,
+                      n_x_sp_eval_types=n_x_sp_eval_types
+                      )
+    else:
+        dont_attack_N_X_SP(args=args,
+                           population=population,
+                           curriculum=curriculum,
+                           unseen_teammates_len=unseen_teammates_len,
+                           n_x_sp_eval_types=n_x_sp_eval_types
+                           )
+
+
+def attack_N_X_SP(args, population, curriculum, unseen_teammates_len, adversary_play_config, attack_rounds, n_x_sp_eval_types):
```

Comment:
Is this to train a new agent against adversaries in the set of teammates? Would it be possible to have one function that does both given a population? Then you could make more diverse population sets instead of adv vs standard?

Reply:
I am not sure if I understand  your question, could you clarify it? The difference between adv and standard is that, adversaries are trained to attack the agent in an iterative manner:
1. train adv0 to target the primary0 agent
2. train primary1 with adv + pop
3. train new adv1 that targets primary1
4. train primary2 with adv1 + adv0 + pop
5. repeat
whereas the standard (i.e the population) could be any agent with any policy that is created once and is not created iteratively

Reply:
So is the goal of this function to train the adversary?

Reply:
yes the latter: train adv + train ego with adv -> repeat for attack_rounds

Reply:
ah, I think the names tripped me up. Maybe something like `joint_adv_N_X_SP` and `N_X_SP` might make sense since attack/dont_attack makes it seem like some sort of post ego-agent training function

Reply:
I agree with Naren that the naming is a little confusing